hazelcast/src/main/java/com/hazelcast/map/impl/query/MapLocalParallerQueryRunner.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.query;
+
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.map.impl.MapServiceContext;
+import com.hazelcast.query.PagingPredicate;
+import com.hazelcast.query.Predicate;
+import com.hazelcast.query.impl.QueryableEntry;
+import com.hazelcast.query.impl.predicates.QueryOptimizer;
+import com.hazelcast.util.executor.ManagedExecutorService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import static com.hazelcast.map.impl.query.MapQueryEngineUtils.waitForResult;
+import static com.hazelcast.query.PagingPredicateAccessor.getNearestAnchorEntry;
+import static com.hazelcast.util.SortingUtil.getSortedSubList;
+
+/**
+ * Specialization of the {@link MapLocalQueryRunner} - the only difference is that
+ * the query evaluation per partition is run in a PARALLEL fashion.
+ *
+ * Runs query operations in the calling thread (thus blucking it)
+ * Query evaluation per partition is run in a PARALLEL fashion.
+ * <p>
+ * Used by query operations only: QueryOperation & QueryPartitionOperation
+ * Should not be used by proxies or any other query related objects.
+ */
+public class MapLocalParallerQueryRunner extends MapLocalQueryRunner {
```

Comment:
After studying the sources a bit longer; I think there are 2 concerns conflated into a single class hierarchy.

You have the concerns:
-  Query Running; so actually chewing through the data with or without index and all that.
-  Query Threading; assigning thread to a query to get processed.

In the OperationService these 2 concerns are separated into 2 classes:
- OperationExecutor
- OperationRunner

Perhaps I'm missing something since processing is quite complex, but would it not make sense to do something similar here.